proxy/proxy.go

```diff
@@ -46,25 +48,28 @@ type Config struct {
 }
 
 type Proxy struct {
+	sync.RWMutex
 	Config
 	client              *docker.Client
 	dockerBridgeIP      string
 	hostnameMatchRegexp *regexp.Regexp
 	weaveWaitVolume     string
+	waitChan            map[string]chan struct{}
 }
 
 func NewProxy(c Config) (*Proxy, error) {
-	p := &Proxy{Config: c}
+	p := &Proxy{Config: c, waitChan: make(map[string]chan struct{})}
 
 	if err := p.TLSConfig.loadCerts(); err != nil {
 		Log.Fatalf("Could not configure tls for proxy: %s", err)
 	}
 
-	client, err := docker.NewClient(dockerSockUnix)
+	client, err := weavedocker.NewClient(dockerSockUnix)
 	if err != nil {
 		return nil, err
 	}
-	p.client = client
+	p.client = client.Client
+	client.AddObserver(p)
```

Comment:
Shouldn't this be done right at the end, so it doesn't call us before `p` is "finished"?